Kaggle - First ML Model


*   Tutorial link:
*   https://www.kaggle.com/code/dansbecker/your-first-machine-learning-model/tutorial
*   We are going to examine the Melborne dataset with this notebook



This script will provide a list of all the columns in the dataset

In [3]:
import pandas as pd

melbourne_data = pd.read_csv('melb_data.csv')
melbourne_data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

We can check the results of our previous script with this one. Both scripts yield 21 columns

In [4]:
df = pd.read_csv('melb_data.csv')
df

Suburb           Address  Rooms Type      Price Method  \
0         Abbotsford      85 Turner St      2    h  1480000.0      S   
1         Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2         Abbotsford      5 Charles St      3    h  1465000.0     SP   
3         Abbotsford  40 Federation La      3    h   850000.0     PI   
4         Abbotsford       55a Park St      4    h  1600000.0     VB   
...              ...               ...    ...  ...        ...    ...   
13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  \
0        Biggin   3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0   
1        Biggin   4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0   
2        Biggin   4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0   
3        Biggin   4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0   
4        Nelson   4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0   
...         ...         ...       ...       ...  ...       ...  ...       ...   
13575     Barry  26/08/2017      16.7    3150.0  ...       2.0  2.0     652.0   
13576  Williams  26/08/2017       6.8    3016.0  ...       2.0  2.0     333.0   
13577     Raine  26/08/2017       6.8    3016.0  ...       2.0  4.0     436.0   
13578   Sweeney  26/08/2017       6.8    3016.0  ...       1.0  5.0     866.0   
13579   Village  26/08/2017       6.3    3013.0  ...       1.0  1.0     362.0   

       BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0               NaN        NaN        Yarra -37.79960   144.99840   
1              79.0     1900.0        Yarra -37.80790   144.99340   
2             150.0     1900.0        Yarra -37.80930   144.99440   
3               NaN        NaN        Yarra -37.79690   144.99690   
4             142.0     2014.0        Yarra -37.80720   144.99410   
...             ...        ...          ...       ...         ...   
13575           NaN     1981.0          NaN -37.90562   145.16761   
13576         133.0     1995.0          NaN -37.85927   144.87904   
13577           NaN     1997.0          NaN -37.85274   144.88738   
13578         157.0     1920.0          NaN -37.85908   144.89299   
13579         112.0     1920.0          NaN -37.81188   144.88449   

                       Regionname Propertycount  
0           Northern Metropolitan        4019.0  
1           Northern Metropolitan        4019.0  
2           Northern Metropolitan        4019.0  
3           Northern Metropolitan        4019.0  
4           Northern Metropolitan        4019.0  
...                           ...           ...  
13575  South-Eastern Metropolitan        7392.0  
13576        Western Metropolitan        6380.0  
13577        Western Metropolitan        6380.0  
13578        Western Metropolitan        6380.0  
13579        Western Metropolitan        6543.0  

[13580 rows x 21 columns]

We can address missing data by dropping missing data using this script


*   Rows that drop off include
*   85 Turner Street
*   40 Federation Lane
*   As a result we go from 13580 rows to 6196 rows



In [7]:
melbourne_data = melbourne_data.dropna(axis=0)
melbourne_data

Suburb          Address  Rooms Type      Price Method  \
1        Abbotsford  25 Bloomburg St      2    h  1035000.0      S   
2        Abbotsford     5 Charles St      3    h  1465000.0     SP   
4        Abbotsford      55a Park St      4    h  1600000.0     VB   
6        Abbotsford     124 Yarra St      3    h  1876000.0      S   
7        Abbotsford    98 Charles St      2    h  1636000.0      S   
...             ...              ...    ...  ...        ...    ...   
12205    Whittlesea    30 Sherwin St      3    h   601000.0      S   
12206  Williamstown      75 Cecil St      3    h  1050000.0     VB   
12207  Williamstown    2/29 Dover Rd      1    u   385000.0     SP   
12209       Windsor  201/152 Peel St      2    u   560000.0     PI   
12212    Yarraville  54 Pentland Pde      6    h  2450000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
1             Biggin   4/02/2016       2.5    3067.0  ...       1.0  0.0   
2             Biggin   4/03/2017       2.5    3067.0  ...       2.0  0.0   
4             Nelson   4/06/2016       2.5    3067.0  ...       1.0  2.0   
6             Nelson   7/05/2016       2.5    3067.0  ...       2.0  0.0   
7             Nelson   8/10/2016       2.5    3067.0  ...       1.0  2.0   
...              ...         ...       ...       ...  ...       ...  ...   
12205            Ray  29/07/2017      35.5    3757.0  ...       2.0  1.0   
12206       Williams  29/07/2017       6.8    3016.0  ...       1.0  0.0   
12207       Williams  29/07/2017       6.8    3016.0  ...       1.0  1.0   
12209  hockingstuart  29/07/2017       4.6    3181.0  ...       1.0  1.0   
12212        Village  29/07/2017       6.3    3013.0  ...       3.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
1         156.0         79.00     1900.0        Yarra -37.80790   144.99340   
2         134.0        150.00     1900.0        Yarra -37.80930   144.99440   
4         120.0        142.00     2014.0        Yarra -37.80720   144.99410   
6         245.0        210.00     1910.0        Yarra -37.80240   144.99930   
7         256.0        107.00     1890.0        Yarra -37.80600   144.99540   
...         ...           ...        ...          ...       ...         ...   
12205     972.0        149.00     1996.0   Whittlesea -37.51232   145.13282   
12206     179.0        115.00     1890.0  Hobsons Bay -37.86558   144.90474   
12207       0.0         35.64     1967.0  Hobsons Bay -37.85588   144.89936   
12209       0.0         61.60     2012.0  Stonnington -37.85581   144.99025   
12212    1087.0        388.50     1920.0  Maribyrnong -37.81038   144.89389   

                  Regionname Propertycount  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
6      Northern Metropolitan        4019.0  
7      Northern Metropolitan        4019.0  
...                      ...           ...  
12205      Northern Victoria        2170.0  
12206   Western Metropolitan        6380.0  
12207   Western Metropolitan        6380.0  
12209  Southern Metropolitan        4380.0  
12212   Western Metropolitan        6543.0  

[6196 rows x 21 columns]

The next script will provide a column that we will use as a prediction target.

In [9]:
y = melbourne_data.Price
y

1        1035000.0
2        1465000.0
4        1600000.0
6        1876000.0
7        1636000.0
           ...    
12205     601000.0
12206    1050000.0
12207     385000.0
12209     560000.0
12212    2450000.0
Name: Price, Length: 6196, dtype: float64

The columns that are inputted into our model (and later used to make predictions) are called "features." In our case, those would be the columns used to determine the home price. Sometimes, you will use all columns except the target as features. Other times you'll be better off with fewer features.

Our next script looks at a few features which are assigned to the value "X"

In [10]:
melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']
X = melbourne_data[melbourne_features]
X

Rooms  Bathroom  Landsize  Lattitude  Longtitude
1          2       1.0     156.0  -37.80790   144.99340
2          3       2.0     134.0  -37.80930   144.99440
4          4       1.0     120.0  -37.80720   144.99410
6          3       2.0     245.0  -37.80240   144.99930
7          2       1.0     256.0  -37.80600   144.99540
...      ...       ...       ...        ...         ...
12205      3       2.0     972.0  -37.51232   145.13282
12206      3       1.0     179.0  -37.86558   144.90474
12207      1       1.0       0.0  -37.85588   144.89936
12209      2       1.0       0.0  -37.85581   144.99025
12212      6       3.0    1087.0  -37.81038   144.89389

[6196 rows x 5 columns]

We can review the data we'll be using to predict house prices using the describe method and the head method, which shows the top few rows.

In [11]:
X.describe()

Rooms     Bathroom      Landsize    Lattitude   Longtitude
count  6196.000000  6196.000000   6196.000000  6196.000000  6196.000000
mean      2.931407     1.576340    471.006940   -37.807904   144.990201
std       0.971079     0.711362    897.449881     0.075850     0.099165
min       1.000000     1.000000      0.000000   -38.164920   144.542370
25%       2.000000     1.000000    152.000000   -37.855438   144.926198
50%       3.000000     1.000000    373.000000   -37.802250   144.995800
75%       4.000000     2.000000    628.000000   -37.758200   145.052700
max       8.000000     8.000000  37000.000000   -37.457090   145.526350

In [12]:
X.head()

Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954

Next we will import the scikit-learn library and use it to define a decision tree model and fit it with features and target variable

In [13]:
from sklearn.tree import DecisionTreeRegressor

# Define model. Specify a number for random_state to ensure same results each run
melbourne_model = DecisionTreeRegressor(random_state=1)

# Fit model
melbourne_model.fit(X, y)

DecisionTreeRegressor(random_state=1)

This allows us to make some predictions

In [14]:
print("Making predictions for the following 5 houses:")
print(X.head())
print("The predictions are")
print(melbourne_model.predict(X.head()))

Making predictions for the following 5 houses:
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954
The predictions are
[1035000. 1465000. 1600000. 1876000. 1636000.]
